c/tskit/trees.c

```diff
@@ -2636,12 +2639,568 @@ check_sites(const tsk_id_t *sites, tsk_size_t num_sites, tsk_size_t num_site_row
+    tsk_bit_array_free(self->node_samples);
+    tsk_safe_free(self->tree);
+    tsk_safe_free(self->node_samples);
+    tsk_safe_free(self->parent);
+    tsk_safe_free(self->edges_out);
+    tsk_safe_free(self->edges_in);
+    tsk_safe_free(self->branch_len);
+}
+
+static int
+advance_collect_edges(iter_state *s, tsk_id_t index)
+{
+    int ret = 0;
+    tsk_id_t j, e;
+    tsk_size_t i;
+    double left, right;
+    tsk_tree_position_t pos;
+    tsk_tree_t *tree = s->tree;
+    const double *restrict edge_left = s->tree->tree_sequence->tables->edges.left;
+    const double *restrict edge_right = s->tree->tree_sequence->tables->edges.right;
+
+    // Either we're seeking forward one step from some nonzero position in the tree, or
+    // from the beginning of the tree sequence.
+    if (tree->index != TSK_NULL || index == 0) {
+        ret = tsk_tree_next(tree);
+        if (ret < 0) {
+            goto out;
+        }
+        pos = tree->tree_pos;
+        i = 0;
+        for (j = pos.out.start; j != pos.out.stop; j++) {
+            s->edges_out[i] = pos.out.order[j];
+            i++;
+        }
+        s->n_edges_out = i;
+        i = 0;
+        for (j = pos.in.start; j != pos.in.stop; j++) {
+            s->edges_in[i] = pos.in.order[j];
+            i++;
+        }
+        s->n_edges_in = i;
+    } else {
+        // Seek from an arbitrary nonzero position from an uninitialized tree.
+        tsk_bug_assert(tree->index == -1);
+        ret = tsk_tree_seek_index(tree, index, 0);
+        if (ret < 0) {
+            goto out;
+        }
+        pos = tree->tree_pos;
+        i = 0;
+        if (pos.direction == TSK_DIR_FORWARD) {
+            left = pos.interval.left;
+            for (j = pos.in.start; j != pos.in.stop; j++) {
+                e = pos.in.order[j];
+                if (edge_left[e] <= left && left < edge_right[e]) {
+                    s->edges_in[i] = pos.in.order[j];
+                    i++;
+                }
+            }
+        } else {
+            right = pos.interval.right;
+            for (j = pos.in.start; j != pos.in.stop; j--) {
+                e = pos.in.order[j];
+                if (edge_right[e] >= right && right > edge_left[e]) {
+                    s->edges_in[i] = pos.in.order[j];
+                    i++;
+                }
+            }
+        }
+        s->n_edges_out = 0;
+        s->n_edges_in = i;
+    }
+    ret = 0;
+out:
+    return ret;
+}
+
+static int
+compute_two_tree_branch_state_update(const tsk_treeseq_t *ts, tsk_id_t c,
+    tsk_bit_array_t *child_samples, const iter_state *A_state, const iter_state *B_state,
+    tsk_size_t state_dim, tsk_size_t result_dim, int sign, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, double *result)
+{
+    int ret = 0;
+    double a_len, b_len;
+    double *restrict B_branch_len = B_state->branch_len;
+    // TODO: is this early return okay?
+    b_len = B_branch_len[c] * sign;
+    if (b_len == 0) {
+        return ret;
```

Comment:
Should test this early return - but it's not clear what it's needed for?  How do you get a branch length of zero, and do you really need to defend against it here?

Reply:
You're right, it's not actually something we would run into. I wasn't able to trigger this conditional. I've removed it.